sling/pyapi/pyarray.cc

```diff
@@ -0,0 +1,200 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sling/pyapi/pyarray.h"
+
+#include "sling/pyapi/pystore.h"
+
+namespace sling {
+
+// Python type declarations.
+PyTypeObject PyArray::type;
+PySequenceMethods PyArray::sequence;
+PyTypeObject PyItems::type;
+
+PyMethodDef PyArray::methods[] = {
+  {"store", (PyCFunction) &PyArray::GetStore, METH_NOARGS, ""},
+  {"data", (PyCFunction) &PyArray::Data, METH_KEYWORDS, ""},
+  {nullptr}
+};
+
+void PyArray::Define(PyObject *module) {
+  InitType(&type, "sling.Array", sizeof(PyArray));
+  type.tp_dealloc = reinterpret_cast<destructor>(&PyArray::Dealloc);
+  type.tp_str = &PyArray::Str;
+  type.tp_iter = &PyArray::Items;
+  type.tp_hash = &PyArray::Hash;
+  type.tp_methods = methods;
+
+  type.tp_as_sequence = &sequence;
+  sequence.sq_length = &PyArray::Size;
+  sequence.sq_item = &PyArray::GetItem;
+  sequence.sq_ass_item = &PyArray::SetItem;
+  sequence.sq_contains = &PyArray::Contains;
+
+  RegisterType(&type);
+}
+
+void PyArray::Init(PyStore *pystore, Handle handle) {
+  // Add array as root object for store to keep it alive in the store.
+  InitRoot(pystore->store, handle);
+
+  // Add reference to store to keep it alive.
+  this->pystore = pystore;
+  Py_INCREF(pystore);
+}
+
+void PyArray::Dealloc() {
+  // Unlock tracking of handle in store.
+  Unlink();
+
+  // Release reference to store.
+  Py_DECREF(pystore);
+}
+
+Py_ssize_t PyArray::Size() {
+  return array()->length();
+}
+
+PyObject *PyArray::GetItem(Py_ssize_t index) {
+  // Check array bounds.
+  ArrayDatum *arr = array();
+  if (index < 0 || index >= arr->length()) {
+    PyErr_SetString(PyExc_IndexError, "Array index out of bounds");
+    return nullptr;
+  }
+
+  // Return array element.
+  return pystore->PyValue(arr->get(index));
+}
+
+int PyArray::SetItem(Py_ssize_t index, PyObject *value) {
+  // Check that array is writable.
+  if (!Writable()) return -1;
+
+  // Check array bounds.
+  if (index < 0 || index >= array()->length()) {
```

Comment:
For extra coolness, can we also support negative indices like Python :) ?

Reply:
Extra coolness added :)